File: LeetCode/Medium/0090_Subsets_2/Solution.cpp
// 2022-05-18
class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        vector<vector<int>> subsets = {{}};
        sort(nums.begin(), nums.end());
        
        int idx = 0;
        while (idx < nums.size()) {
            int cur = nums[idx];
            vector<vector<int>> oldSubsets = subsets;           
            while (idx < nums.size() && nums[idx] == cur) {
                vector<vector<int>> newSubsets;
                for (vector<int> set : oldSubsets) {
                    vector<int> newSubset = set;
                    newSubset.push_back(cur);
                    newSubsets.push_back(newSubset);
                }
                subsets.insert(subsets.end(), newSubsets.begin(), newSubsets.end());
                oldSubsets = move(newSubsets);
                idx++;
            }
        }
        return subsets;
    }
};

File: LeetCode/Medium/0090_Subsets_2/Solution.py
from typing import *

## 2022-11-09 (iterative)
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        counts = dict()
        for num in nums:
            counts[num] = counts.get(num, 0)+1
        
        subsets = [[]]
        for num in range(-10, 11):
            n = len(subsets)
            for _ in range(counts.get(num, 0)):
                additional = []
                for k in range(1, n+1):
                    additional.append(subsets[-k] + [num])
                subsets += additional
        
        return subsets
